File: idscp2/src/main/java/de/fhg/aisec/ids/idscp2/drivers/default_driver_impl/rat/tpm2d/TPM2dVerifierConfig.java
package de.fhg.aisec.ids.idscp2.drivers.default_driver_impl.rat.tpm2d;

import de.fhg.aisec.ids.idscp2.messages.Tpm2dAttestation.IdsAttestationType;
import org.checkerframework.checker.nullness.qual.NonNull;

import java.net.URI;
import java.net.URISyntaxException;
import java.security.cert.Certificate;

/**
 * A configuration class for TPM2d RatVerifier Driver
 *
 * @author Leon Beckmann (leon.beckmann@aisec.fraunhofer.de)
 */
public class TPM2dVerifierConfig {

    private Certificate localCertificate;
    @NonNull
    private URI ttpUri;
    @NonNull
    private IdsAttestationType expectedAType = IdsAttestationType.BASIC;
    private int expectedAttestationMask = 0;

    private TPM2dVerifierConfig() {
        try {
            ttpUri = new URI("https://invalid-ttp-uri/rat-verify");
        } catch (URISyntaxException e) {
            throw new RuntimeException(e);
        }
    }

    public static class Builder {
        private static final TPM2dVerifierConfig config = new TPM2dVerifierConfig();

        @NonNull
        public Builder setTtpUri(URI ttpUri) {
            config.ttpUri = ttpUri;
            return this;
        }

        @NonNull
        public Builder setLocalCertificate(Certificate localCert) {
            config.localCertificate = localCert;
            return this;
        }

        @NonNull
        public Builder setExpectedAttestationType(IdsAttestationType aType) {
            config.expectedAType = aType;
            return this;
        }

        public Builder setExpectedAttestationMask(int mask) {
            config.expectedAttestationMask = mask;
            return this;
        }

        public TPM2dVerifierConfig build() {
            return config;
        }
    }

    @NonNull
    public URI getTtpUri() {
        return ttpUri;
    }

    @NonNull
    public IdsAttestationType getExpectedAType() {
        return expectedAType;
    }

    public Certificate getLocalCertificate() {
        return localCertificate;
    }

    public int getExpectedAttestationMask() {
        return expectedAttestationMask;
    }
}

File: idscp2/src/main/kotlin/de/fhg/aisec/ids/idscp2/default_drivers/rat/tpm2d/TPM2dVerifierConfig.kt
package de.fhg.aisec.ids.idscp2.default_drivers.rat.tpm2d

import de.fhg.aisec.ids.idscp2.messages.Tpm2dAttestation
import java.net.URI
import java.net.URISyntaxException
import java.security.cert.Certificate

/**
 * A configuration class for TPM2d RatVerifier Driver
 *
 * @author Leon Beckmann (leon.beckmann@aisec.fraunhofer.de)
 */
class TPM2dVerifierConfig private constructor() {
    var localCertificate: Certificate? = null
        private set
    var ttpUri: URI? = null
        private set
    var expectedAType = Tpm2dAttestation.IdsAttestationType.BASIC
        private set
    var expectedAttestationMask = 0
        private set

    class Builder {
        fun setTtpUri(ttpUri: URI): Builder {
            config.ttpUri = ttpUri
            return this
        }

        fun setLocalCertificate(localCert: Certificate): Builder {
            config.localCertificate = localCert
            return this
        }

        fun setExpectedAttestationType(aType: Tpm2dAttestation.IdsAttestationType): Builder {
            config.expectedAType = aType
            return this
        }

        fun setExpectedAttestationMask(mask: Int): Builder {
            config.expectedAttestationMask = mask
            return this
        }

        fun build(): TPM2dVerifierConfig {
            return config
        }

        companion object {
            private val config = TPM2dVerifierConfig()
        }
    }

    init {
        ttpUri = try {
            URI("https://invalid-ttp-uri/rat-verify")
        } catch (e: URISyntaxException) {
            throw RuntimeException(e)
        }
    }
}